src/mumble/AudioInput.h

```diff
@@ -29,6 +32,57 @@ struct OpusEncoder;
 struct DenoiseState;
 typedef boost::shared_ptr<AudioInput> AudioInputPtr;
 
+struct AudioChunk
+{
+	AudioChunk()                       : mic(nullptr), spk(nullptr) {}
+	explicit AudioChunk(short *mic)    : mic(mic),     spk(nullptr) {}
+	AudioChunk(short *mic, short *spk) : mic(mic),     spk(spk)     {}
+	bool empty() const { return mic == nullptr; }
+
+	short *mic;
+	short *spk;
+};
+
+/*
+ * According to https://www.speex.org/docs/manual/speex-manual/node7.html
+ * "It is important that, at any time, any echo that is present in the input
+ * has already been sent to the echo canceller as echo_frame."
+ * Thus, we artificially introduce a small lag in the microphone by means of
+ * a queue, so as to be sure the speaker data always precedes the microphone.
+ *
+ * There are conflicting requirements for the queue:
+ * - it has to be small enough not to cause a noticeable lag in the voice
+ * - it has to be large enough not to force us to drop packets frequently
+ *   when the addMic and addEcho callbacks are called in a jittery way
+ * - its fill level must be controlled so it does not operate towards zero
+ *   elements size, as this would not provide the lag required for the
+ *   echo canceller to work properly.
+ *
+ * The current implementation uses a 5 elements queue, with a control
+ * statemachine that introduces packet drops to control the fill level
+ * to at least 2 (plus or minus one) and less than 4 elements.
+ * With a 10ms chunk, this queue should introduce a ~20ms lag to the voice.
+ */
+class Resynchronizer
+{
+public:
+	void addMic(short *mic);
+	AudioChunk addSpk(short *spk);
```

Comment:
I think I'd prefer to not abbreviate "speaker" in the function name
```suggestion
	AudioChunk addSpeaker(short *spk);
```